### PR TITLE
Implement the Importer Service

### PR DESCRIPTION
**Pull request by edith007**
**What this PR does**:

**Which issue(s) this PR fixes**:
Fixes https://github.com/cortexproject/cortex/issues/4956

To provide clarity and transparency to the development process, we have broken down the remaining tasks for this PR into the following list:

- [x] Create the import API endpoint
- [x] Download the compressed Prometheus TSDB block from the given URL
- [x] Validate the compressed block
- [ ] Upload the uncompressed block to the long-term storage
- [ ] Create a CLI tool for the importer service

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by edith007** [comment]
@alanprot I have implemented the logic to fetch the TSDB blocks from any given url... I have assumed that our `importer/import` API will have a parameter called `url` which will specify the url from where the TSDB block can be downloaded... I am making a get request to the URL... upon receiving the response of our request to fetch the TSDB block, I am placing the TSDB block in /tmp/tsdb_blocks/ folder...

**Comment by alanprot** [comment]
hi @edith007 .. this is a very good start.

Looking at https://github.com/cortexproject/cortex/pull/5229/files seems that we agreed to send to no send the URL anymore but the block bundle itself compressed (gziped)?

We also need after that to decompress, validate and upload to the long term storage right?

**Comment by edith007** [comment]
Hi @alanprot, Just wanted to update you on the latest progress on the Cortex importer/import API. I have successfully implemented the logic to fetch the TSDB block from a given URL, uncompress and validate it, and store it in the local system in the /tmp/tsdb_blocks/ directory.

However, we still need to work on the remaining part, which is to save the block in Cortex. I will be focusing on this next step and keep you posted on the progress.

File: pkg/importer/importer.go [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+}
+
+func downloadPrometheusBlock(url string) (string, error) {
+	// Create the /tmp directory if it doesn't already exist
+	if _, err := os.Stat("/tmp/tsdb_blocks"); os.IsNotExist(err) {
+		err = os.Mkdir("/tmp/tsdb_blocks", 0755)
+		if err != nil {
+			return "", err
+		}
+	}
```
Comment (alanprot) [inline pkg/importer/importer.go:40]: Should we get the `path` of this folder via config? And also this folder should be created "per tenant" and with a random prefix (so we can have multiples requests being processed at the same time?

I see something like: `/data/{tenantId}/{random}`

File: pkg/importer/importer.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+	i.Service = services.NewBasicService(i.starting, i.running, i.stopping)
+	return i
+}
+
+func (i *Importer) SampleRPC(ctx context.Context, request *importerpb.SampleRequest) (*importerpb.SampleResponse, error) {
+	//TODO implement me
+	panic("implement me")
+}
+
+func downloadPrometheusBlock(url string) (string, error) {
```
Comment (alanprot) [inline pkg/importer/importer.go:33]: where this function is being used?

**Review by alanprot — COMMENTED** [review]
I did not see where you are uploading the block to the long term storage. Did i miss something:

I think you can use this code as reference: https://github.com/thanos-io/thanos/blob/4d6a1d7e011a2921e6d5aec2b9189ca008830ae9/pkg/compact/compact.go#L960

**Comment by edith007** [comment]
@alanprot I was thinking that we can start the import of the TSDB block to the cortex from here itself (once we have the block on the machine) ... but I am not sure how to do that (was exploring that) ... is there some API in cortex ingester that takes  the url of the long term storage tsdb block and ingests it into cortex ??

another question... why do we want to upload to the long term storage? to have backups in case the machine on which the code is running goes down??

**Pull request closed without merging**